### PR TITLE
perf: Optimize CometSparkToColumnar for columnar input

### LINKED ISSUE #798: Implement more efficient conversion from Spark column to Comet/Arrow column
### What is the problem the feature request solves?

In `CometRowToColumnarExec` (soon to be renamed to `CometSparkToColumnarExec`) we have code for converting from Spark columns to Comet columns, but it it does this by first converting the Spark columns to rows and then converts the rows to Comet columns.

### Describe the potential solution

It would be better to implement direct Spark column to Comet column conversion

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mbutrovich**
## Which issue does this PR close?

Closes #798 .

## Rationale for this change

CometSparkToColumnar currently iterates over input batches using a row iterator, regardless of whether the data is columnar or not.

## What changes are included in this PR?

- Extended `ArrowWriter` to write an entire column at a time, with an optimization if the column does not contain any NULL values.
- `ArrowBatchIterator` made into an abstract class `SparkToArrowConverter` with specialized subclasses depending on the source batch type (`ArrowBatchIteratorFromColumnBatch`, `ArrowBatchIteratorFromInternalRows`). I welcome naming feedback.
- `CometSparkToColumnarExec` now uses one of these specialized iterators depending on the child's output.

## How are these changes tested?

There is a new test to exercise various types coming out of the Spark Parquet reader (BatchScan), and a new config in CometExecSuite (`SQL Parquet - Comet (Scan disabled, Exec)`) to test this configuration. On the main branch, the benchmark results are:
```
Project + Filter Exec (0.0% zeros):        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                   57             73          13        183.4           5.5       1.0X
SQL Parquet - Comet (Scan)                            53             67          10        198.4           5.0       1.1X
SQL Parquet - Comet (Scan, Exec)                      52             64           9        201.4           5.0       1.1X
SQL Parquet - Comet (Scan disabled, Exec)             97            109           8        108.6           9.2       0.6X

Project + Filter Exec (50.0% zeros):       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                   42             49           8        250.2           4.0       1.0X
SQL Parquet - Comet (Scan)                            41             54           9        254.2           3.9       1.0X
SQL Parquet - Comet (Scan, Exec)                      45             59          10        234.8           4.3       0.9X
SQL Parquet - Comet (Scan disabled, Exec)             85             98          12        123.3           8.1       0.5X

Project + Filter Exec (95.0% zeros):       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                   27             36          10        384.7           2.6       1.0X
SQL Parquet - Comet (Scan)                            28             46          12        372.0           2.7       1.0X
SQL Parquet - Comet (Scan, Exec)                      33             53          12        319.5           3.1       0.8X
SQL Parquet - Comet (Scan disabled, Exec)             73             86          12        143.4           7.0       0.4X
```
With the changes in this PR:
```
Project + Filter Exec (0.0% zeros):        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                   63             78          12        165.3           6.1       1.0X
SQL Parquet - Comet (Scan)                            54             72          12        195.2           5.1       1.2X
SQL Parquet - Comet (Scan, Exec)                      54             66          11        193.9           5.2       1.2X
SQL Parquet - Comet (Scan disabled, Exec)             79             94          11        133.2           7.5       0.8X

Project + Filter Exec (50.0% zeros):       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                   43             58          13        246.1           4.1       1.0X
SQL Parquet - Comet (Scan)                            42             60          14        250.3           4.0       1.0X
SQL Parquet - Comet (Scan, Exec)                      46             66          15        229.0           4.4       0.9X
SQL Parquet - Comet (Scan disabled, Exec)             72             94          15        145.5           6.9       0.6X

Project + Filter Exec (95.0% zeros):       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark                                   27             40          14        383.7           2.6       1.0X
SQL Parquet - Comet (Scan)                            28             47          14        376.2           2.7       1.0X
SQL Parquet - Comet (Scan, Exec)                      34             56          15        312.5           3.2       0.8X
SQL Parquet - Comet (Scan disabled, Exec)             63             87          18        166.2           6.0       0.4X
```

### CODE REVIEW

**Comment by andygrove** [comment]
I see one test failure:

```
2024-08-30T01:50:56.4878673Z - SparkToColumnar over BatchScan (Spark Parquet reader) *** FAILED *** (7 seconds, 294 milliseconds)
2024-08-30T01:50:56.4880378Z   leaves.forall(((x$20: org.apache.spark.sql.execution.SparkPlan) => x$20.supportsColumnar)) was false (CometExecSuite.scala:1587)
```

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala
Comment (andygrove) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala:161]: This comment isn't related to any changes in this PR, but I was a little surprised at our current api where we call `setNull()` with no arguments and it knows to set null at the current index, but when we call `setValue` we pass the index. It just seems inconsistent. We can maybe revisit this design in the future.

File: spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala:1586]: very minor nit: rather than using strings to represent booleans in `Seq("true", "false")`, I think it would be better to have booleans here so that we can just have have `if (parquetVectorized)` here and when we need to access it as a string when setting configs we can call `.toString`.

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala
Comment (mbutrovich) [reply common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala:161]: Cruft from the original implementation from Spark's `ArrowWriter`, it seems.

File: spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala
Comment (mbutrovich) [reply spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala]: Should I file an issue to propagate this across the codebase in a future PR?

Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala]: I don't have a strong opinion either way. I suppose it could be a good first issue for someone and it would be nice to clean up existing tests.

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @mbutrovich! It would be good to get a review from @viirya as well before merging.

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala
Comment (viirya) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala:111]: Shouldn't it be?
```suggestion
    count = count + input.numElements()
```

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala
Comment (viirya) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala:116]: ditto
```suggestion
    count = count + input.numElements()
```

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala
Comment (viirya) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala:49]:
```suggestion
  abstract private[sql] class ArrowBatchIterBase(
```

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala
Comment (viirya) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala:104]:
```suggestion
  private[sql] class RowToArrowBatchIter(
```

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala
Comment (viirya) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala:150]:
```suggestion
  private[sql] class ColumnBatchToArrowBatchIter(
```

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala
Comment (viirya) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala:136]:
```suggestion
  def rowToArrowBatchIter(
```

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala
Comment (viirya) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala:195]:
```suggestion
  def columnarBatchToArrowBatchIter(
```

**Review by viirya — APPROVED** [review]

**Review by viirya — COMMENTED** [review]
Looks good. Some naming suggestions.

File: common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala
Comment (viirya) [reply common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/ArrowWriters.scala:111]: Oh, you always write a ColumnarArray into ArrowWriter once then reset it.

**Pull request merged**
